### ISSUE TITLE
'Find in Files' does not search through files with no extension

### ISSUE THREAD

**Issue by schadjo** [notepad-plus-plus/notepad-plus-plus#14617]
### Description of the Issue

When in Find in Files dialog, if your filter is `<something>*.`, clicking Find All will not search through any files.

The Search Results shows `Search "whatever" (0 hits in 0 files of 40 searched)` for the above search.

### Steps to Reproduce the Issue

1.  Make a folder with some files named, say, `O9000.`  and `O9800.`, and put some searchable string in them, say, `XXX`.
2.  Open the Find in Files dialog.
3.  In the Filters field, type `O*.` .
4.  In the Find what field, type `XXX`.
5.  Set the Directory field to wherever you put `O9000.` and `O9800.`.
6.  Hit Find All.
7.  Search results will show no hits because 0 files were searched.

### Expected Behavior

I expect the search function to return the search string in `O*.` files when the search string exists in `O9000.` and `O9800.`. 

### Actual Behavior

No matches are returned.

### Debug Information

Notepad++ v8.5.4   (64-bit)
Build time : Jun 17 2023 - 20:42:45
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : "C:\PROGRA~3\FANUC\FANUCL~1\Tmp\jschad\WFLAD3\MAP.TXT"
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Pro (64-bit)
OS Version : 22H2
OS Build : 19045.3930
Current ANSI codepage : 1252
Plugins : 
    ComparePlugin (2.0.2)
    GcodeHelper (4.3)
    mimeTools (2.9)
    NppConverter (4.5)
    NppExec (0.8)
    NppExport (0.4)
    XMLTools (3.1.1.13)

**Comment by molsonkiko** [comment]
I'm not surprised you're having this issue: __files with an empty extension shouldn't exist, and most applications won't even create them.__

TLDR: if files with that name shouldn't exist, you should consider fixing whatever is creating them (or begging the creator to fix it), not asking Notepad++ to handle your corner case.

I tried making a file with an empty extension using all of the following:
1. Creating one using the GUI in the File Explorer
2. Saving one in Notepad++
3. Creating one with Python.

In every one of those cases, __what I got instead was a file with no extension, not a file with an empty extension.__

**Comment by alankilborn** [comment]
> files with an empty extension shouldn't exist

I think OP means "no extension".
In other words, I don't know what an "empty extension" would be, other than no extension.

You can create an extensionless file in Explorer, from the command line, and many other ways.
In Explorer and Cmd, you can either use the `.` or not, e.g. `echo x>test` or `echo x>test.` ... both yield the same result, a file with the exact name `test` in the file system.

**Comment by schadjo** [comment]
@alankilborn Yes, you are right. My mistake.

I work with a lot of FANUC CNCs which strip any file extensions from the gcode files placed on the machine. When retrieved for editing, they have no extension. 

Being able to search though a bunch of no-extension files in multiple subdirectories would be a win for me.

**Comment by molsonkiko** [comment]
@schadjo OK, thanks for clarifying what you meant. Yes, I see files with *no extension* all the time, so your use case makes a lot more sense now.

It turns out you can do this! You just need to familiarize yourself with globbing, which is discussed to some extent [here](https://npp-user-manual.org/docs/searching/). A more comprehensive guide could be found by googling `glob syntax`.

What you want to do is put the filter `!*.?*` (which excludes all files with at least one character in their extension) in the `Filters` box.

If you want to restrict which files without extensions you look at, you can just add those filters separately. So `**2 !*.?*` would search for all extensionless files in all subdirectories with names that end with the digit `2`

**Comment by alankilborn** [comment]
So I think the answer here is RTFM (Read The Fine Manual) and this issue may now be closed.

**Comment by schadjo** [comment]
I see what you're saying, but for 30+ years I've been able to do from a DOS/command prompt dir/s *. and list every file and directory with no extension. It seems counterintuitive to look for *. and not find *. when they exist.

**Comment by ioc2e3** [comment]
> 違反直覺

agree ...

( The file exists ) but ( cannot be found ) is really counterintuitive ...

.

**Comment by molsonkiko** [comment]
So after a bit of fooling around, I've figured out the following:

Suppose you have a file named `new2` (no extension). Then:
- `*.*` __matches__
- `*` __matches__
- `**` __matches__ (when in recursive mode)
- `w*` __does not match__
- `*2` __matches__

I tested this on Notepad++ 8.5.8, 8.6.4, and 8.5.1, and I'm pretty sure the behavior is consistent.

So in short, it appears that find in files usually works fine for files with no extension, but this is one corner case where it's broken. Of course, this is still a bug, and I'm going to keep looking into it.

**Comment by alankilborn** [comment]
Is a possible solution as "simple" as:

if a user's filter spec (or an individual part of it) ends with `.`, remove it and let the search for matching files proceed; then examine results and if a matched filename has a `.` in it, drop that file from consideration for the content-search phase

**Comment by alankilborn** [comment]
`::FindFirstFile` and `::FindNextFile` functions seem to be doing what is desired when searching for e.g. `*.`.
( In fact, for experimenting with these in an easy manner, I found [THIS](https://stackoverflow.com/questions/66968972/python-ctypes-api-call-results-in-access-violation) really cool Python program from a StackOverflow answer. )

After a little bit more looking, the problem appears to be in the use of the `::PathMatchSpec` Win API function.

---

@molsonkiko said:

>...I'm going to keep looking into it

If you don't mind, I'd like to try coming up with a fix for this issue.
But, if you really want to yourself, let me know and I'll cease, as you expressed interest in this issue first.

**Comment by alankilborn** [comment]
Similar complaint from 2017:

https://community.notepad-plus-plus.org/topic/13196/find-in-files-search-only-files-with-no-extension/

**Comment by alankilborn** [comment]
Side note:  The user manual even [mentions](https://npp-user-manual.org/docs/searching/#:~:text=note%20that%20the-,PathMatchSpec,-()%20Windows%20API%20is) `::PathMatchSpec`!

**Comment by alankilborn** [comment]
In addition to extensionless files that should be matched with `*.`, directory exclusion of this nature `!\*.` and `!+\*.` should exclude extensionless directories (currently it does NOT exclude them).

**Comment by alankilborn** [comment]
Just a note :

In CMD, because it was compared to existing Notepad++ filter capability earlier https://github.com/notepad-plus-plus/notepad-plus-plus/issues/14617#issuecomment-1899529574 :

* a `dir *.` will match files to _do_ contain a `.` (at the start) such as `.gitignore`, `.editorconfig`, etc., as well as other files/dirs that don't contain `.` (i.e., extensionless files)
* `dir .*` will match only files that start with a `.`
* it doesn't seem possible for `dir` to return only files that have no `.` whatsoever (but maybe some better CMD guru than me knows how to make it do this??)

A possible future version of Notepad++ _could_ match files that have no `.` whatsoever, via this filterspec: `*. !.*`

**Comment by alankilborn** [comment]
So I came up with some new code for N++ that will allow `*.` to work.
However, it is somewhat of a performance killer regarding time it takes to search.

<details>
  <summary>Click to view search scenario and timing results details</summary>

```txt
Searched N++ 8.6.4 source code tree
Find what: \z
Search mode: Regular expression
Filters: *

Roughly 2050 files are hit.

Timing with N++ 8.6.4: ~5s
Timing with new code that supports *. : ~12s
```
</details>

I will continue to look at possibilities to both support `*.` and retain similar performance to existing N++ code.
I think it might be possible with a lot of restructuring of how locating the fileset to search is done (which is probably a deal-breaker for such code being accepted into N++ -- too much risk).
